src/exampleTypes/CheckableResultTable.js

```diff
@@ -2,50 +2,49 @@ import React from 'react'
 import _ from 'lodash/fp'
 import { observer } from 'mobx-react'
 import F from 'futil-js'
-import InjectTreeNode from '../utils/injectTreeNode'
 import { getResults, getRecord } from '../utils/schema'
+import { len } from '../utils/futil'
+
+let getLabelComponent = ({ Checkbox, selected, getValue }) =>
+  observer(({ node }) => {
+    let results = getResults(node).slice()
+    let allChecked = len(results) === len(F.view(selected))
+    let checkAll = F.sets(
+      allChecked
+        ? []
+        : _.map(
+            _.flow(
+              getRecord,
+              _.iteratee(getValue)
+            ),
+            results
+          ),
+      selected
+    )
+    return <Checkbox checked={allChecked} onChange={checkAll} />
+  })
 
 // Extends ResultTable with a checkbox column
 // Writes to a lens called `selected`, using getValue to map the selected record to a value.
 // getValues uses _.iteratee, so it defaults to identity and supports things like strings to get props
-let CheckableResultTable = InjectTreeNode(
-  observer(
-    ({ node, fields, selected, getValue, Checkbox, ResultTable, ...props }) => {
-      let results = getResults(node).slice()
-      let allChecked = results.length === F.view(selected).length
-      let checkAll = F.sets(
-        allChecked
-          ? []
-          : _.map(
-              _.flow(
-                getRecord,
-                _.iteratee(getValue)
-              ),
-              results
-            ),
-        selected
-      )
-      return (
-        <ResultTable
-          fields={{
-            _checkbox: {
-              hideMenu: true,
-              label: <Checkbox checked={allChecked} onChange={checkAll} />,
-              display: (x, y) => (
-                <Checkbox
-                  {...F.domLens.checkboxValues(
-                    _.iteratee(getValue)(y),
-                    selected
-                  )}
-                />
-              ),
-            },
-            ...fields,
-          }}
-          {...props}
-        />
-      )
-    }
+let CheckableResultTable = observer(
+  ({ fields, selected, getValue, Checkbox, ResultTable, ...props }) => (
+    <ResultTable
+      fields={{
+        _checkbox: {
+          hideMenu: true,
+          label: '',
+          LabelComponent: getLabelComponent({ Checkbox, selected, getValue }),
```

Comment:
You wouldn't need this additional API if you were just able to pass node to the label, which you could do if you leverage injectTreeNode on on CheckableResultTable like it was before. Also - I thought we were going to hold off on performance related things that complicate the code until we did profiling since we don't know objectively how much of a difference it makes. Did you start profiling stuff?

Reply:
I did this because the component was re-rendering too many times in my slow network, so it appeared as it was never going to render.

Reply:
This additional API is just to pass `Checkbox`, `selected` and `getValue`, not the node. Would `injectTreeNode` help to pass `Checkbox`, `selected` and `getValue`?

Reply:
If you had node in scope, you could just have a regular component on label instead of needing to introduce labelComponent

Reply:
@daedalus28 not really, if you pass that regular component, it will be rendered at the point of declaration instead of being rendered as part of the table header, so it will still re-render the whole table.